src/doc/src/reference/build-scripts.md

```diff
@@ -250,6 +254,25 @@ identifier, the value should be a string.
 [conditional compilation]: ../../reference/conditional-compilation.md
 [option-cfg]: ../../rustc/command-line-arguments.md#option-cfg
 
+### `cargo::rustc-check-cfg=CHECK_CFG` {#rustc-check-cfg}
+
+The `rustc-check-cfg` instruction tells Cargo to pass the given expected
+configs to the [`--check-cfg` flag][option-check-cfg] to the compiler. This is
+used for compile-time detection of unexpected conditional compilation configs.
+
+You can use the instruction like this:
+
+```rust,no_run
+// build.rs
+println!("cargo::rustc-check-cfg=cfg(foo)");
+```
+
+Note that it is expected to **always** pass the full list of custom configs
+used in the package. The recommended way to achieve this is put all the
+check-cfg instructions at the top of your `build.rs`.
```

Comment:
What do you mean by "full"?

Reply:
The complete list of all the custom cfgs and their values, even if some combinations are not technically possible.

Reply:
maybe this could be reworded to "all possible `cfg`, regardless of what is currently active".

Also a bit unsure of the idea of "put all at the top".  I expect there would be cases where that'd be a terrible thing to do from a maintenance perspective.

Reply:
> The "put all at the top" is mainly to suggest users to put all of the check-cfg instructions at the start of the build.rs main, as a way to make sure they are before anything else (in particular any conditions or whatever).

The `check-cfg` logic has a logical dependency on the `cfg` logic for generating the correct set of cfgs.  The further away they are, the more likely they are to get out of sync.  We should not be encouraging them to be further away, making them more likely to get out of sync.

(this is assuming multiple `check-cfg` directives can be given)

Reply:
I'm basing this recommendation on the fact that [`bootstrap`](https://github.com/rust-lang/rust/blob/d255c6a57c393db6221b1ff700daea478436f1cd/src/bootstrap/src/lib.rs#L74-L89) and [`libc`](https://github.com/rust-lang/libc/blob/a0f5b4b21391252fe38b2df9310dc65e37b07d9f/build.rs#L5-L30) both defines their custom check-cfgs in a list at the top of `build.rs` (or equivalent). It IMO makes it clear from the start what are the expected cfgs, even if the actual Cargo instructions is berried away.

> (this is assuming multiple check-cfg directives can be given)

yes they can

Reply:
I agree that it's not perfect, but given that we don't have a "support crate", it's the simplest recommendation I can think of.

Maybe we could recommend somethings similar to the `libc` setup, but I fear it will be too involved for users to just apply it and they would roll out a worst version of both recommendation.

Reply:
I agree that having a helper library could make this simpler to verify their own output and I agree it would be too much to tell people to be like libc.

I disagree with keeping this suggestion though especially because we don't expect people to be like libc.

Reply:
I'm fine removing it but not without an alternative.

Since your worried about the instruction being too far apart, what about something like this?

> One way to achieve this, is to group the check-cfg and cfg instructions as much as
possible.
>
> <details>
> 
> ```rust,no_run
> // build.rs
> println!("cargo::rustc-check-cfg=cfg(foo, values(\"bar\"))");
> if foo_bar_condition {
>     println!("cargo::rustc-cfg=foo=\"bar\"");
> }
> ```
> 
> </details>

The example makes it clear about what needs to be done, can be generalized if desired and respect your desired to put them as close as possible.

Reply:
This works.  I'd  also be fine with just not giving direction on this.